network/framework/src/application/storage.rs

```diff
@@ -188,14 +197,31 @@ impl PeersAndMetadata {
             .and_modify(|peer_metadata| {
                 peer_metadata.connection_metadata = connection_metadata.clone()
             })
-            .or_insert_with(|| PeerMetadata::new(connection_metadata));
+            .or_insert_with(|| PeerMetadata::new(connection_metadata.clone()));
 
         // Update the cached peers and metadata
         self.set_cached_peers_and_metadata(peers_and_metadata.clone());
 
+        let net_context = self.get_network_context(&peer_network_id.network_id());
+        let event = ConnectionNotification::NewPeer(connection_metadata, net_context);
+        self.broadcast(event);
+
         Ok(())
     }
 
+    pub fn set_network_contexts(&self, contexts: BTreeMap<NetworkId, NetworkContext>) {
+        self.contexts.store(Arc::new(contexts));
+    }
+
+    // Return NetworkContext for a NetworkId, or fake-up something best effort because it's just for logging.
+    fn get_network_context(&self, network_id: &NetworkId) -> NetworkContext {
+        self.contexts
+            .load()
+            .get(network_id)
+            .cloned()
+            .unwrap_or_else(|| NetworkContext::new(RoleType::Validator, *network_id, PeerId::ZERO))
```

Comment:
Hmm... It might be cleaner to make this return `None` as opposed to an arbitrary context 😄

Reply:
I'd also be fine with plain `unwrap()` and panic, because that should never happen.